File: Editorial/1625_Lexicographically_Smallest_String_After_Applying_Operations/C/1_Approach_Enumeration.c
char* findLexSmallestString(char* s, int a, int b) {
    int n = strlen(s);
    int vis[n];
    memset(vis, 0, sizeof(vis));
    char* res = (char*)calloc(sizeof(char), n + 1);
    strcpy(res, s);
    // double the length of s for convenience in extracting the rotated string t
    char str[2 * n + 1];
    sprintf(str, "%s%s", s, s);
    for (int i = 0; vis[i] == 0; i = (i + b) % n) {
        vis[i] = 1;
        for (int j = 0; j < 10; j++) {
            int k_limit = b % 2 == 0 ? 0 : 9;
            for (int k = 0; k <= k_limit; k++) {
                // before each accumulation, re-truncate t
                char t[n + 1];
                strncpy(t, str + i, n);
                t[n] = '\0';
                for (int p = 1; p < n; p += 2) {
                    t[p] = '0' + (t[p] - '0' + j * a) % 10;
                }
                for (int p = 0; p < n; p += 2) {
                    t[p] = '0' + (t[p] - '0' + k * a) % 10;
                }
                if (strcmp(t, res) < 0) {
                    strcpy(res, t);
                }
            }
        }
    }
    return res;
}

File: Editorial/1625_Lexicographically_Smallest_String_After_Applying_Operations/TypeScript/1_Approach_Enumeration.ts
function findLexSmallestString(s: string, a: number, b: number): string {
    const n = s.length;
    const vis = new Array(n).fill(false);
    let res = s;
    // double the length of s for convenience in extracting the rotated string t
    s = s + s;
    for (let i = 0; !vis[i]; i = (i + b) % n) {
        vis[i] = true;
        for (let j = 0; j < 10; j++) {
            let kLimit = b % 2 === 0 ? 0 : 9;
            for (let k = 0; k <= kLimit; k++) {
                // before each accumulation, re-truncate t
                const t = [...s.slice(i, i + n)];
                for (let p = 1; p < n; p += 2) {
                    t[p] = String.fromCharCode(
                        "0".charCodeAt(0) +
                            ((t[p].charCodeAt(0) - "0".charCodeAt(0) + j * a) %
                                10),
                    );
                }
                for (let p = 0; p < n; p += 2) {
                    t[p] = String.fromCharCode(
                        "0".charCodeAt(0) +
                            ((t[p].charCodeAt(0) - "0".charCodeAt(0) + k * a) %
                                10),
                    );
                }
                const tStr = t.join("");
                if (tStr < res) {
                    res = tStr;
                }
            }
        }
    }
    return res;
}
